### ISSUE TITLE
TEE_OUT_OF_MEMORY when loading TA (Hikey960 - AOSP)

### ISSUE THREAD

**Issue by bsupiot** [OP-TEE/optee_os#2597]
Hi,

I'm trying to run the hello_world example on the Hikey960.( I have the same issue in the xtest suit. )

When I run hello_world I have random result, rarely the hello_world work fine otherwise I have the following error : 

> optee_example_hello_world: TEEC_OpenSession failed with code 0xffff0008 origin 0x3

So it's the TEE_ERROR_ITEM_NOT_FOUND. I go deeper in the code and observe execution flow. I arrive when optee want to load the TA. 

1. Optee try to load the TA as pseudo trusted app -> TEE_ERROR_ITEM_NOT_FOUND -> It's normal, hello_world isn't a pta.
2. So optee try to load the TA as user TA looking up in the Secure Storage -> TEE_ERROR_ITEM_NOT_FOUND -> It's normal, hello_world isn't in Secure Storage
3. Finally optee try to load the TA as user TA looking up in the REE. So RPC is executed to find the TA in REE file system and the TA is founded.

So the problem doesn't seems to be TEE_ERROR_ITEM_NOT_FOUND. I continue to investigate and place a log in the following condition:

https://github.com/OP-TEE/optee_os/blob/9551f4e55c19cba96b1187b82dcea9bd2512f758/core/arch/arm/kernel/ree_fs_ta.c#L62

When I have the issue I saw my log, the problem seems to be TEE_ERROR_OUT_OF_MEMORY prompt by https://github.com/OP-TEE/optee_os/blob/9551f4e55c19cba96b1187b82dcea9bd2512f758/core/arch/arm/kernel/ree_fs_ta.c#L61

I find the source of the TEE_ERROR_OUT_OF_MEMORY : https://github.com/OP-TEE/optee_os/blob/9551f4e55c19cba96b1187b82dcea9bd2512f758/core/arch/arm/mm/mobj.c#L480

The function core_pbuf_is() return false and then goto err.

I don't understand why this issue happen. Someone can help me ? 

Thanks in advance

**Comment by jenswi-linaro** [comment]
In explanation could be that OP-TEE OS in secure world isn't aware of all the non-secure memory used by normal world. Perhaps you need to recompile with `CFG_DRAM_SIZE_GB=4`?

**Comment by bsupiot** [comment]
Thanks for your fast answer. Need I change some address in config files ? When my board boot I have the following issue : 

> NOTICE:  BL2: v1.5(release):v1.5-649-g7e8a891
NOTICE:  BL2: Built : 15:08:17, Oct 10 2018
NOTICE:  ufs: H**** device must set VS_DebugSaveConfigTime 0x10
NOTICE:  BL2: Booting BL31
NOTICE:  BL31: v1.5(release):v1.5-649-g7e8a891
NOTICE:  BL31: Built : 15:08:17, Oct 10 2018
D/TC:0 0 add_phys_mem:539 VCORE_UNPG_RX_PA type TEE_RAM_RX 0x3f000000 size 0x00079000
D/TC:0 0 add_phys_mem:539 VCORE_UNPG_RW_PA type TEE_RAM_RW 0x3f079000 size 0x00187000
D/TC:0 0 add_phys_mem:539 TA_RAM_START type TA_RAM 0x3f200000 size 0x00e00000
D/TC:0 0 add_phys_mem:539 TEE_SHMEM_START type NSEC_SHM 0x3ee00000 size 0x00200000
D/TC:0 0 add_phys_mem:539 CONSOLE_UART_BASE type IO_NSEC 0xffe00000 size 0x00200000
D/TC:0 0 verify_special_mem_areas:483 SDP memory [3e800000 3ec00000]
D/TC:0 0 verify_special_mem_areas:483 NSEC DDR memory [0 3e000000]
D/TC:0 0 verify_special_mem_areas:483 NSEC DDR memory [40000000 100000000]
E/TC:0 0 verify_special_mem_areas:514 [40000000 100000000] intersecs [ffe00000 100000000]
E/TC:0 0 Panic 'Special memory intersection' at core/arch/arm/mm/core_mmu.c:515 <verify_special_mem_areas>
E/TC:0 0 Call stack:
E/TC:0 0  0x000000003f0143e4
D/TC:0 0 unwind_stack_arm64:56 FP out of bounds 0x3f079b80

**Comment by jenswi-linaro** [comment]
@bsupiot, thanks for the information. It seems we have a bug here, do you agree @jforissier?

**Comment by jforissier** [comment]
@jenswi-linaro, yes, DRAM1_SEC_SIZE is certainly wrong, and in fact I don't think 0xBFE00000 is correct either: the [SoC reference manual](https://github.com/96boards/documentation/raw/master/consumer/hikey/hikey960/hardware-docs/HiKey960_SoC_Reference_Manual.pdf) page 2-23 says that the DRAM address range is 0x00000000 - 0xDFFFFFFF for a total of 3.5GB, so the limit would seem to be 0xE0000000, not 0xFFE00000.

**Comment by bsupiot** [comment]
@vchong I will ask my manager if I can share my manifest, however I have some issues with my integration : 
* No SePolicy integration
* USB3.0 doen't worked (so can't use GUI)
* Instructions in init.common.rc aren't execute (so I need to start tee-supplicant manually, create directories, ect)
* When I run xtest I have some TA panics (I don't know if it's normal) . Xtest Result :+1: 

> 16065 subtests of which 3 failed
75 test cases of which 1 failed
0 test case was skipped
TEE test application done!

So It's good for experimental TA

**Comment by vchong** [comment]
A big thanks to @bsupiot who has kindly shared his implementations here: https://github.com/bsupiot/hikey960-aosp-optee/wiki! For those who are interested, please give it a go and contribute back patches if possible. Thanks!
